examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/ntp.c

```diff
@@ -0,0 +1,301 @@
+#include "wifi_module.h"
+#include "wifi_globals.h"
+
+#include "ntp.h"
+
+int32_t sntp_current_time = 0;
+sntp_response* last_sntp_response = NULL;
+
+/**
+   * @brief  free the space used by an sntp_response
+   * @param  *sntp_server is a char array with the relevant URL
+   * @retval WiFi_Status_t: WiFi_MODULE_SUCCESS on success, see wifi_interface.h
+   */
+void sntp_free_response( sntp_response* r )
+{
+    if( NULL != r->response )
+    {
+        free(r->response);
+    }
+    free(r);
+}
+
+/**
+   * @brief  Builds a char array with a basic SNTP request to be sent via UDP
+   *         Packet description:
+   *           - Flags 1 byte
+   *               * Leap: 3 bits
+   *               * Version: 3 bits
+   *               * Mode: 2 bits
+   *           - Stratum 1 byte
+   *           - Polling 1 byte
+   *           - Precision 1 byte
+   *           - Root Delay 4 bytes
+   *           - Root Dispersion 4 bytes
+   *           - Reference Identifier 4 bytes
+   *           - Reference Timestamp 8 bytes
+   *           - Origin Timestamp 8 bytes
+   *           - Receive Timestamp 8 bytes
+   *           - Transmit Timestamp 8 bytes
+   * @param  request_buf is a pointer to a pre-allocated 48 byte char array
+   * @retval void
+   */
+void sntp_build_request( char* request_buf )
+{
+    char sntp_pkt[SNTP_MSG_SIZE] = {
+    0xe3, 0x00, 0x03, 0xfa, 0x00, 0x01, 0x00, 0x00,
+    0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0xd5, 0x22, 0x0e, 0x35, 0xb8, 0x76, 0xab, 0xea};
+    memset(request_buf, 0, SNTP_MSG_SIZE);
+    memcpy(request_buf, sntp_pkt, SNTP_MSG_SIZE);
+}
+
+/**
+   * @brief  Create a UDP socket to the SNTP server, connect to it and return
+   *         its ID by setting *sock_id
+   * @param  *sntp_server is a char array with the relevant URL
+   *         sntp_port is the port to be used for SNTP communication
+   *         *sock_id will be set to the appropriate socket ID
+   * @retval WiFi_Status_t: WiFi_MODULE_SUCCESS on success, see wifi_interface.h
+   */
+WiFi_Status_t sntp_connect( char* sntp_server, uint32_t sntp_port, uint8_t* sock_id )
+{
+    WiFi_Status_t status = WiFi_MODULE_SUCCESS;
+    uint8_t sntp_protocol = 'u'; //UDP
+    printf("\r\n\tStablishing UDP connection to SNTP server %s:%lu",
+           sntp_server,
+           sntp_port);
+    status = wifi_socket_client_open((uint8_t*)sntp_server, sntp_port,
+                                     &sntp_protocol, sock_id);
+
+    return status;
+}
+
+/**
+   * @brief
+   * @param  sock_id is the socket ID set by sntp_connect() in stnp_start()
+   * @retval WiFi_Status_t: WiFi_MODULE_SUCCESS on success, see wifi_interface.h
+   */
+WiFi_Status_t sntp_disconnect( uint8_t sock_id )
+{
+    WiFi_Status_t status = WiFi_MODULE_SUCCESS;
+    printf("\r\n>>Closing UDP connection to SNTP server...");
```

Comment:
Yeah, exactly this, the programmer can add only this comment, but there is no such thing: "UDP connection" so cannot close it. But it can stay here because the Wifi API dictates this.